examples/iiqka_moveit_task_constructor_example/kuka_moveit_task_constructor/src/mtc_depalletizing_task_node.cpp

```diff
@@ -0,0 +1,314 @@
+        move_pose.pose.orientation.w = 0.0;
+        move_stage->setGoal(move_pose);
+        task.add(std::move(move_stage));
+
+        // Define Place Stage
+        auto place_stage = std::make_unique<mtc::stages::MoveRelative>(
+          "place_" + std::to_string(4 * k + 2 * j + i), cartesian_planner);
+        place_stage->properties().set("marker_ns", "place_" + std::to_string(4 * k + 2 * j + i));
+        place_stage->properties().set("link", eef_frame);
+        place_stage->properties().configureInitFrom(mtc::Stage::PARENT, {"group"});
+        geometry_msgs::msg::Vector3Stamped vec_place;
+        vec_place.header.frame_id = "world";
+        vec_place.vector.z = -(2 - k) * PALLET_DISTANCE - 0.02;
+        place_stage->setDirection(vec_place);
+        task.add(std::move(place_stage));
+
+        // Detach Object Stage
+        auto detach_stage = std::make_unique<mtc::stages::ModifyPlanningScene>(
+          "detach object_" + std::to_string(4 * k + 2 * j + i));
+        detach_stage->detachObject("pallet_" + std::to_string(4 * k + 2 * j + i), eef_frame);
+        task.add(std::move(detach_stage));
+
+        // Define Lift2 Stage
+        auto lift2_stage = std::make_unique<mtc::stages::MoveRelative>(
+          "lift2_" + std::to_string(4 * k + 2 * j + i), cartesian_planner);
+        lift2_stage->properties().set("marker_ns", "lift2_" + std::to_string(4 * k + 2 * j + i));
+        lift2_stage->properties().set("link", eef_frame);
+        lift2_stage->properties().configureInitFrom(mtc::Stage::PARENT, {"group"});
+        geometry_msgs::msg::Vector3Stamped vec_lift2;
+        vec_lift2.header.frame_id = "world";
+        vec_lift2.vector.z = (2 - k) * PALLET_DISTANCE + 0.02;
+        lift2_stage->setDirection(vec_lift2);
+        task.add(std::move(lift2_stage));
+      }
+    }
+  }
+
+  return task;
+}
+
+bool MTCDepalletizingTaskNode::doTask(mtc::Task & task)
+{
+  if (task.stages()->numChildren() == 0)
+  {
+    RCLCPP_ERROR_STREAM(LOGGER, "No task set.");
+    return false;
+  }
+
+  try
+  {
+    task.init();
+  }
+  catch (mtc::InitStageException & e)
+  {
+    RCLCPP_ERROR_STREAM(LOGGER, e);
+    return false;
+  }
+
+  if (!task.plan(1000))
+  {
+    RCLCPP_ERROR_STREAM(LOGGER, "Task planning failed");
+    return false;
+  }
+  task.introspection().publishSolution(*task.solutions().front());
+
+  auto result = task.execute(*task.solutions().front());
+  if (result.val != moveit_msgs::msg::MoveItErrorCodes::SUCCESS)
+  {
+    RCLCPP_ERROR_STREAM(LOGGER, "Task execution failed");
+    return false;
+  }
+
+  return true;
+}
+
+int main(int argc, char ** argv)
+{
+  rclcpp::init(argc, argv);
+
+  rclcpp::NodeOptions options;
+  options.automatically_declare_parameters_from_overrides(true);
+
+  auto mtc_depalletizing_task_node = std::make_shared<MTCDepalletizingTaskNode>(options);
+  rclcpp::executors::MultiThreadedExecutor executor;
+
+  auto spin_thread = std::make_unique<std::thread>(
```

Comment:
is there a reason for making this thread a unique ptr instead of simple var?

Reply:
Some people like it, some does not. I sometimes use pointers for threading (mostly when they might throw an exception).
Advantages:
Automatic Cleanup: std::unique_ptr ensures the thread is cleaned up when it goes out of scope, preventing resource leaks.
Exception Safety: Ensures the thread’s destructor is called even if an exception is thrown, joining the thread if necessary.
Clear Ownership: Indicates single ownership, making the code easier to understand and maintain.

Disadvantages:
Overhead: May introduce unnecessary overhead in terms of performance and complexity.
Move Semantics: std::thread is already move-only, and wrapping it in a smart pointer can complicate the design.
Readability: Can reduce code readability and clarity by adding another layer of abstraction

Yes, I know it's a simple demo application and std::thread is pretty robust itself, I can change it to a simple var if you want. :)

Reply:
the custom deleter is a better approach, as it can gracefully stop a thread, while in the case of a smart pointer, if the pointer runs out of scope, the thread's destructor is called, which terminates the thread abruptly. That's why I manually joined the thread in the example. Having a custom deleter would be more ideal, I just did not think this simple example would need it.

Reply:
But compared to the simple std::thread, what is the difference without the custom deleter? the simple var would also run out of scope and the destructor would be called. So I do not see why the unique ptr is better in this case, as without the deleter it will terminate the same way (also in case of an exception, where join() would not be called automatically)

(I do not necessarily want to change to simple var, just want to understand the effects of both and without a custom deleter I think your first two advantages are not valid, as the behaviours of both cases are identical)

Reply:
I read a bit about how exactly it works. Yes, you are right, the only reason one might want to use a smart pointer for storing a thread is for ownership purposes. It does not really apply for our case, so changing it back to a simple var makes a lot of sense.